publisher-sdk/src/main/java/com/criteo/publisher/csm/MetricSendingQueueConsumer.java

```diff
@@ -100,22 +105,50 @@ public void runSafely() throws IOException {
         return;
       }
 
-      boolean success = false;
+      Collection<Metric> metricsToRollback = new ArrayList<>(metrics);
       try {
-        MetricRequest request = createRequest(metrics);
-        api.postCsm(request);
-        success = true;
+        Map<MetricRequest, Collection<Metric>> requests = createRequests(metrics);
+        for (Entry<MetricRequest, Collection<Metric>> entry : requests.entrySet()) {
+          api.postCsm(entry.getKey());
+          metricsToRollback.removeAll(entry.getValue());
+        }
       } finally {
-        if (!success) {
-          rollback(metrics);
+        if (!metricsToRollback.isEmpty()) {
+          rollback(metricsToRollback);
         }
       }
     }
 
-    private MetricRequest createRequest(Collection<Metric> metrics) {
+    private Map<MetricRequest, Collection<Metric>> createRequests(Collection<Metric> metrics) {
+      Map<Integer, Collection<Metric>> metricsPerProfile = new LinkedHashMap<>();
+      for (Metric metric : metrics) {
+        Integer profileId = metric.getProfileId();
+        if (profileId == null) {
+          profileId = Integration.FALLBACK.getProfileId();
+        }
+
+        Collection<Metric> metricsForProfile = metricsPerProfile.get(profileId);
+        if (metricsForProfile == null) {
+          metricsForProfile = new ArrayList<>();
+          metricsPerProfile.put(profileId, metricsForProfile);
+        }
+        metricsForProfile.add(metric);
+      }
+
       String sdkVersion = buildConfigWrapper.getSdkVersion();
-      int profileId = buildConfigWrapper.getProfileId();
-      return MetricRequest.create(metrics, sdkVersion, profileId);
+      Map<MetricRequest, Collection<Metric>> metricRequests = new LinkedHashMap<>();
+      for (Entry<Integer, Collection<Metric>> entry : metricsPerProfile.entrySet()) {
+        int profileId = entry.getKey();
+        Collection<Metric> metricsForProfile = entry.getValue();
+        MetricRequest metricRequest = MetricRequest.create(
+            metricsForProfile,
+            sdkVersion,
+            profileId
+        );
+        metricRequests.put(metricRequest, metricsForProfile);
+      }
```

Comment:
Would it make sense that `MetricRequest` keep & expose its underlying metrics, as it takes `metricsForProfile` as input, rather than this `Map<MetricRequest, Collection<Metric>`?

Reply:
The `MetricRequest` is a DTO object autogenerated from an interface and JSON schema is based directly on its properties. I would avoid putting too much logic in this class.

Also, here the `metricForProfile` is kept only to handle the partial rollback.